inst/include/dust/dust.hpp

```diff
@@ -405,26 +427,117 @@ class Dust {
     if (n_pars_ == 0) {
       // One parameter set; shuffle among all particles
       const size_t np = particles_.size();
-      real_t u = dust::unif_rand(rng_.state(0));
+      real_t u = dust::unif_rand(rng_.state(n_particles_total_));
       dust::filter::resample_weight(it_weights, np, u, 0, it_index);
     } else {
       // Multiple parameter set; shuffle within each group
       // independently (and therefore in parallel)
       const size_t np = particles_.size() / n_pars_;
+      std::vector<real_t> u;
+      for (size_t i = 0; i < n_pars_; ++i) {
+        u.push_back(dust::unif_rand(rng_.state(n_particles_total_)));
+      }
 #ifdef _OPENMP
       #pragma omp parallel for schedule(static) num_threads(n_threads_)
 #endif
       for (size_t i = 0; i < n_pars_; ++i) {
         const size_t j = i * np;
-        real_t u = dust::unif_rand(rng_.state(j));
-        dust::filter::resample_weight(it_weights + j, np, u, j, it_index + j);
+        dust::filter::resample_weight(it_weights + j, np, u[i], j, it_index + j);
       }
     }
 
     reorder(index);
     stale_device_ = true;
   }
 
+  void resample_device(dust::device_array<real_t>& weights,
```

Comment:
This method is pretty big - can the implementation be farmed off into a free function?